pkg/syncer/namespace/namespace_process.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+)
+
+func (c *Controller) process(ctx context.Context, key string) error {
+	_, clusterAwareName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("Invalid key %q: %w", key, err)
+		return nil
+	}
+	clusterName, namespaceName := clusters.SplitClusterAwareKey(clusterAwareName)
+	namespaceGVR := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "namespaces"}
+
+	// If the clusterName is empty, the enqueued namespace is a downstream namespace,
+	// the downstream namespace needs to be deleted if there's no upstream namespace anymore.
+	//
+	// This happens on syncer start or during a resync.
+	if clusterName.Empty() {
+		_, exists, err := c.getUpstreamNSForDownstreamNS(namespaceName)
```

Comment:
Split this into
1. Get downstream namespace
2. Return early if it's not the relevant namespace
3. Check if upstream namespace exists or not

Reply:
updated, PTAL